test/compilable/b12647.d

```diff
@@ -0,0 +1,9 @@
+void assertThrown12647(E)(lazy E expression)
+{
+    expression();
+}
+
+void test12647() nothrow pure @safe @nogc
+{
+    assertThrown12647(0);
+}
```

Comment:
Alright, thanks for the thorough explanation, I'm not sure if it is possible to turn the errors into warnings in a relatively painless way.
Just pushed a working draft, gotta adjust the tests when I have some time.